### PR TITLE
first pass migrate graph api to subfolder

### PR DESCRIPTION
**Pull request by ryanfkeepers**
## Description

In order to establish a standard api around our
graph client usage, and thus be able to mock
for testing, we need to migrate graph client
usage into another pacakge.  This is the first
step in that process of refactoring.

All changes are code relocation importing and
exporting may change as needed.  No logic
was altered.

## Does this PR need a docs update or release note?

- [x] :no_entry: No 

## Type of change

- [x] :robot: Test
- [x] :hamster: Trivial/Minor

## Issue(s)

* #1967

## Test Plan

- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

File: src/internal/connector/exchange/api/api.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/microsoft/kiota-abstractions-go/serialization"
+
+	"github.com/alcionai/corso/src/internal/connector/graph"
+)
+
+// ---------------------------------------------------------------------------
+// common types
+// ---------------------------------------------------------------------------
+
+// DeltaUpdate carries details about delta retrieval in aggregators
```
Comment (dadams39) [inline src/internal/connector/exchange/api/api.go:15]: Probably needs a bit more information

File: src/internal/connector/exchange/api/contacts.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+			builder = users.NewItemContactFoldersItemContactsDeltaRequestBuilder(*nextLink, gs.Adapter())
+		}
+
+		return nil
+	}
+
+	if len(oldDelta) > 0 {
+		err := getIDs(users.NewItemContactFoldersItemContactsDeltaRequestBuilder(oldDelta, gs.Adapter()))
+		// happy path
```
Comment (dadams39) [inline src/internal/connector/exchange/api/contacts.go:178]: Could we expand upon this comment?

**Review by dadams39 — APPROVED** [review]
a few nits. Linting and other tests will have to pass prior to merging.

**Comment by ashmrtn** [comment]
What's the end goal for the `api` struct? Is it that it's going to implement all functions for all data categories in exchange and those will be the interface?
For example, is the end goal for it to be like
```go
type GraphAPI interface {
  GetMailFolders(...) (folders, error)
  CreateMailFolder(...) error
  CreateMailFolderWithParent(...) error
  GetMessages(...) (messages, error)
  ...

  GetContactFolders(...) (folders, error)
  CreateContactFolder(...) error
  CreateContactFolderWithParent(...) error
  GetContacts(...) (messages, error)
  ...

  // Events functions
}
```

The above seems like it's going to end up as a pretty big interface, which makes mocking annoying because all functions need to be implemented. It may be easier in the long run to instead split things by data category and have multiple implementations of a single interface like
```go
type GraphAPI interface {
  GetFolders(...) (folders, error)
  CreateFolder(...) error
  CreateNestedFolder(...) error
  GetItemsInFolder(...) (items, error)
}

type MailAPI struct {
  ...
}

// Implement interface functions
```

Despite the fact that each implementation works with different data types we're already on the way to wrapping those returned types with our own abstractions that hide the differences. All of the data types also essentially have the same execution flow for backing stuff up or restoring it (may also want to split backup and restore into different interfaces). For backup it's:
1. enumerate folders
2. get item IDs for each folder
3. get item data for each item ID

for restore it's:
1. make folder(s) for restored items
2. restore items in each folder

**Comment by ryanfkeepers** [comment]
> It may be easier in the long run to instead split things by data category

@ashmrtn You're getting a little ahead of the progress here.  Two PRs down the line or so we do exactly this.

**Pull request merged**